core/platform/email/mailgun_email_services_test.py

```diff
@@ -53,3 +58,129 @@ def test_sending_email(self):
             mailgun_email_services.send_mail(
                 feconf.SYSTEM_EMAIL_ADDRESS, feconf.ADMIN_EMAIL_ADDRESS,
                 'subject', 'body', 'html', bcc_admin=False)
+
+
+    def test_sending_email_data(self):
+        """Verifies that the data sent in send_mail is correct."""
+        mailgun_api = self.swap(feconf, 'MAILGUN_API_KEY', 'api')
+        mailgun_domain = self.swap(feconf, 'MAILGUN_DOMAIN_NAME', 'domain')
+        allow_emailing = self.swap(feconf, 'CAN_SEND_EMAILS', True)
+
+        # This data should have been sent in the requests.post().
+        expected = {'from': feconf.SYSTEM_EMAIL_ADDRESS, 'to':
+                    feconf.ADMIN_EMAIL_ADDRESS, 'subject': 'subject',
+                    'text': 'body', 'html': 'html'}
+
+        # Lambda function, will replace requests.post() in send_mail.
+        req_post_lambda = (lambda domain_name, auth=None, data=None:
```

Comment:
I don't understand this. Also, I haven't seen this pattern elsewhere. Could you explain what it does, and how it is better than sending the request, and then asserting the dict is correct?

Reply:
I swapped out the requests.post() in send_mail since the only thing that mattered was asserting that the correct data was being passed into the function. I don't know what you mean by "asserting the dict is correct" -- replacing the request lambda is how I'm getting the dictionary passed into the requests.post() function, and I do assert it later in the function. Is there another way to do this? The preexisting code was already using self.swap() to test exceptions, so I assumed self.swap() would be the ideal way to test the actual emailing functionality, since I don't believe I have any access to the data sent through the requests library.

Reply:
Also, self.swap() accepts a lambda function. That is why I created that lambda. I tried creating a separate helper function (using **def**), but self.swap did not like that, so I am assuming that it only accepts lambdas.

Reply:
Ah, It does accept normal functions. I am not familiar with working of lambdas. Fine to leave it as is, as it looks like it works correctly.  :)